File: src/main/java/com/nearvanilla/icecream/commands/ModuleCommand.java
/* Licensed under GNU Affero General Public License v3.0 */
package com.nearvanilla.icecream.commands;

import com.nearvanilla.icecream.IceCream;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;

public class ModuleCommand implements CommandExecutor {

  @Override
  public boolean onCommand(
      @NotNull CommandSender commandSender,
      @NotNull Command command,
      @NotNull String s,
      @NotNull String[] strings) {
    // Check if player is executing command.
    if (!(commandSender instanceof Player player)) {
      commandSender.sendMessage("Only players can use this command.");
      return true;
    }
    // Check if 2 arguments are provided.
    if (strings.length == 2) {
      player.sendMessage(
          Component.text("Usage: /module <enable|disable> <module>", NamedTextColor.RED));
    }
    String enableDisable = strings[0];
    String module = strings[1];
    // Check if enable or disable is provided.
    if (!enableDisable.equals("enable") && !enableDisable.equals("disable")) {
      player.sendMessage(
          Component.text("Usage: /module <enable|disable> <module>", NamedTextColor.RED));
      return true;
    }
    // Check if module is provided.
    if (module.isEmpty()) {
      player.sendMessage(
          Component.text("Usage: /module <enable|disable> <module>", NamedTextColor.RED));
      return true;
    }
    // Switch statement to enable or disable module.
    switch (module) {
      case "lightning":
        if (enableDisable.equals("enable")) {
          IceCream.lightningModule.enable();
          player.sendMessage(Component.text("Lightning module enabled.", NamedTextColor.GREEN));
        } else {
          IceCream.lightningModule.disable();
          player.sendMessage(Component.text("Lightning module disabled.", NamedTextColor.GREEN));
        }
        break;
      case "worldtour":
        if (enableDisable.equals("enable")) {
          IceCream.worldTourModule.enable();
          player.sendMessage(Component.text("WorldTour module enabled.", NamedTextColor.GREEN));
        } else {
          IceCream.worldTourModule.disable();
          player.sendMessage(Component.text("WorldTour module disabled.", NamedTextColor.GREEN));
        }
        break;
      default:
        player.sendMessage(
            Component.text("Usage: /module <enable|disable> <module>", NamedTextColor.RED));
        break;
    }
    return true;
  }
}

File: src/main/java/com/nearvanilla/iceCream/commands/ModuleCommand.kt
/* Licensed under the GNU Affero General Public License. */
package com.nearvanilla.iceCream.commands

import com.nearvanilla.iceCream.IceCream
import io.papermc.paper.command.brigadier.CommandSourceStack
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.NamedTextColor
import org.bukkit.entity.Player
import org.incendo.cloud.annotations.Argument
import org.incendo.cloud.annotations.Command
import org.incendo.cloud.annotations.CommandDescription
import org.incendo.cloud.annotations.Permission

class ModuleCommand {
    private val enabledComponent = Component.text("Module enabled.", NamedTextColor.GREEN)

    private val disableComponent = Component.text("Module disabled.", NamedTextColor.RED)

    private val invalidActionComponent =
        Component.text("Invalid action. Expecting either 'enable' or 'disable'.", NamedTextColor.RED)

    private val invalidModuleComponent = Component.text(
        "Invalid Module. Please check the documentation for a list of modules and try again.",
        NamedTextColor.RED
    )

    @Command("module <action> <module>")
    @CommandDescription("Enable or disable a module.")
    @Permission("icecream.module")
    fun moduleCommand(
        sourceStack: CommandSourceStack,
        @Argument(value = "action") action: String,
        @Argument(value = "module") module: String?
    ) {
        val sender = sourceStack.sender
        if (sender !is Player) {
            sender.sendMessage("This command can only be run by a player.")
            return
        }
        when (module) {
            "lightning" -> {
                run {
                    if (action == "enable") {
                        IceCream.lightningModule.register()
                        sender.sendMessage(enabledComponent)
                    } else if (action == "disable") {
                        IceCream.lightningModule.unregister()
                        sender.sendMessage(disableComponent)
                    } else {
                        sender.sendMessage(invalidActionComponent)
                    }
                }
                sender.sendMessage(invalidModuleComponent)
            }

            else -> sender.sendMessage(invalidModuleComponent)
        }
    }
}
